lib/std/zig/render.zig

```diff
@@ -2377,6 +2381,24 @@ fn renderSpace(ais: *Ais, tree: Ast, token_index: Ast.TokenIndex, lexeme_len: us
     }
 }
 
+// If the symbol is escaped with @"" syntax, only render it that way if it contains
+// invalid characters or is a keyword, otherwise render it with canonical syntax.
+fn renderCanonicalSymbol(ais: *Ais, lexeme: []const u8) Error!void {
+    if (lexeme[0] == '@') {
+        assert(lexeme.len > 2 and lexeme[1] == '"' and lexeme[lexeme.len - 1] == '"');
+
+        const quoted_contents = lexeme[2 .. lexeme.len - 1];
+        if (std.zig.isValidId(quoted_contents)) {
```

Comment:
Unfortunately this isn't sufficient to handle the example of `@"\x66\x6f\x6f"` given in the linked issue as it does not interpret escape sequences before checking `isValidId()`. `@"\x66\x6f\x6f"` should be canonicalized to `foo`.

Reply:
It's possible in theory to handle this without allocation by combining the logic of `isValidId()` and `parseAppend()` into a single function with a fixed sized buffer large enough to hold the longest keyword.

Alternatively, it wouldn't be too bad to allocate for this if we reused the same buffer for each identifier. This would look something like a scratch `ArrayList(u8)` buffer that is created at the start of `renderTree()` and is only deinitialized when that function is completed. Every `renderCanonicalSymbol()` call would then parse the contents of `@""` quoted identifiers into it, check if the parsed result is a valid id, etc. Then the buffer would be cleared without freeing the memory so that the next `renderCanonicalSymbol()` call may avoid allocating at all if the buffer is large enough.

Reply:
Good ideas. After writing down my "Behavior of canonicalization" comment, I can visualize a two-pass approach that uses a local `var buf: [length_of_longest_keyword_or_primitive]u8`, which I agree is less nasty than option 2.